packages/langium/src/workspace/file-system-provider.ts

```diff
@@ -0,0 +1,53 @@
+/******************************************************************************
+ * Copyright 2022 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import { URI } from 'vscode-uri';
+
+export interface FileSystemNode {
+    readonly isFile: boolean;
+    readonly isDirectory: boolean;
+    readonly name: string;
+    readonly container: URI;
+}
+
+export interface FileSystemFilter {
+    (node: FileSystemNode): boolean;
+}
+
+export interface FileSystemProvider {
+    /**
+     * Reads a document synchronously from a given URI.
+     * @returns The string content of the file with the specified URI.
+     */
+    readFile(uri: URI): Promise<string>;
+    /**
+     * Reads a document asynchronously from a given URI.
+     * @returns The string content of the file with the specified URI.
+     */
+    readFileSync(uri: URI): string;
+    /**
+     * Traverses a directory structure and returns all paths within.
+     * @param root The root path where the traversal starts.
+     * @param filter An optional filter that determines whether a directory or file should be included while traversing.
+     * @returns All URIs which match the specified filter in the directory structure.
+     */
+    traverse(root: URI, filter?: FileSystemFilter): Promise<URI[]>;
+}
+
+export class NullFileSystemProvider implements FileSystemProvider {
```

Comment:
What is this good for? Langium will not work when this service implementation is active.